crates/astria-sequencer-relayer/src/relayer.rs

```diff
@@ -161,21 +344,129 @@ impl Relayer {
     ///
     /// `Relayer::run` never returns an error. The return type is
     /// only set to `eyre::Result` for convenient use in `SequencerRelayer`.
-    pub(crate) async fn run(self) -> eyre::Result<()> {
-        let mut interval = time::interval(self.sequencer_poll_period);
-        loop {
-            interval.tick().await;
-            match self.get_and_submit_latest_block().await {
-                Err(e) => warn!(error = ?e, "failed to get latest block from sequencer"),
-                Ok(new_state) if new_state != *self.state_tx.borrow() => {
-                    _ = self.state_tx.send_replace(new_state);
+    #[instrument(name = "Relayer::run", skip_all)]
+    pub(crate) async fn run(mut self) -> eyre::Result<()> {
+        let wait_for_da = self.wait_for_data_availability_layer(5, Duration::from_secs(5), 2.0);
+        let wait_for_seq = self.wait_for_sequencer(5, Duration::from_secs(5), 2.0);
+        match tokio::try_join!(wait_for_da, wait_for_seq) {
+            Ok(((), ())) => {}
+            Err(err) => return Err(err).wrap_err("failed to start relayer"),
+        }
+        // .await
+        // .wrap_err("failed establishing connection to data availability layer")?;
+        self.wait_for_sequencer(5, Duration::from_secs(5), 2.0)
+            .await
+            .wrap_err("failed establishing connection to the sequencer")?;
+
+        let mut sequencer_interval = interval(self.sequencer_poll_period);
+        // the first tick resolves immediately, so do it now because we don't
+        // want to poll right after start.
+        sequencer_interval.tick().await;
```

Comment:
Huh, I did this because I was fixing the (mocked) gossip node subscribing to the "blocks" topic too late (so that sequencer-relayer did not have anywhere to send the data).

Now that relayer is waiting for both celestia + at least 1 subscriber to be present this might not be necessary anymore.

Thanks for raising this! I will have to look into it.

Reply:
It turns out this was indeed no longer necessary.